File: app/src/main/java/com/fafu/app/dfb/http/RetrofitFactory.java
package com.fafu.app.dfb.http;

import java.util.Map;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import retrofit2.Retrofit;

public class RetrofitFactory {

    private static OkHttpClient client = new OkHttpClient.Builder()
            .addInterceptor(new HeaderInterceptor())
            .addInterceptor(new CookieInterceptor())
            .build();

    private static Retrofit retrofit = new Retrofit.Builder()
            .client(client)
            .baseUrl("http://cardapp.fafu.edu.cn:8088")
            .build();

    public static <T> T obtainService(Class<T> clazz, Map<String, String> headers) {
        if (headers != null) {
            client.newBuilder()
                    .addInterceptor(chain -> {
                        Request.Builder builder = chain.request().newBuilder();
                        headers.forEach(builder::addHeader);
                        return chain.proceed(builder.build());
                    });
        }
        return retrofit.create(clazz);
    }

}

File: app/src/main/java/com/fafu/app/dfb/http/RetrofitFactory.kt
package com.fafu.app.dfb.http

import okhttp3.OkHttpClient
import retrofit2.Retrofit

object RetrofitFactory {

    private val client = OkHttpClient.Builder()
            .addInterceptor(CookiesInterceptor())
            .addInterceptor(HeaderInterceptor())
            .build()

    private val retrofit = Retrofit.Builder()
            .client(client)
            .baseUrl("http://cardapp.fafu.edu.cn:8088")
            .build()

    fun <T> obtainService(clazz: Class<T>, headers: Map<String, String>? = null): T {
        if (headers != null) {
            client.newBuilder()
                    .addInterceptor {
                        val builder = it.request().newBuilder()
                        headers.forEach { (t, u) ->
                            builder.addHeader(t, u)
                        }
                        it.proceed(builder.build())
                    }
        }
        return retrofit.create(clazz)
    }

}